### ISSUE TITLE
Unable to send command to Itach Ethernet

### ISSUE THREAD

**Issue by sydneyjd** [bitfocus/companion-module-globalcache-itac-ir#4]
Hello,
 I have been trying to use this plugin with a Itach Ethernet, currently configured with a static IP (192.168.1.207). The ILearn program running on the same machine works perfectly for both recieving IR commands from the remote and sending that command to the Itach for triggering my test device.
 Companion is a fresh install of the latest version from the website(2.1.2) on Windows 10, with the ip2ir instance set for the correct IP and showing OK.
 I configure a button with the ip2ir: Chose Port on the down action, ms = 0,  Port = 1 and the appropriate Global Cache formatted IR string in the IR Code box. The only oddity is that If I remove the `sendir,1:1,1` section from the ir code as per the instructions, the IR Code Box will NOT accept the string. It I leave it in it will accept the string. 
 When I hit "test button" in Companion the LAN activity light on the Itach Ethernet flashes in correlation with the test button presses, but nothing happens. No errors in the Companion error or debug log either.

 Am I missing something? Thank you!

**Comment by ethan-tripp** [comment]
Having same problem with Companion version 2.1.2, trying to send commands to IP2IR device. IR TCP commands send fine using the iLearn app.

**Comment by krocheck** [comment]
@JeffreyDavidsz @jpooley-hearst  thoughts?

**Comment by JeffreyDavidsz** [comment]
Nope no experience with this. Seems connection is fine, something seems to be wrong with the sending string

**Comment by jpooley-hearst** [comment]
Hey all, 

I don't think it's actually working as I thought it was, sorry for the trouble. I've had issues setting up a development environment. 

Our issue is likely an extra  `+ ','` on line 129 of index.js

I also neglected to update the documentation... you'd have to remove sendir,1:x (x being 1, 2, or 3 referring to the port number) but no more. The previous instance of this module assumed that the number after that was 1 which is not always the case. 

Here's example code in Python that actually works: 

`s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    s.connect((ip, 4998))
    s.send("sendir,1:1,1,38186,3,1,343,171,19,85,19,85,19,171,19,171,19,85,19,85,19,85,19,85,19,85,19,85,19,85,19,85,19,85,19,85,19,171,19,85,19,1350,343,86,19,3363,343,86,19,3818\r\n".encode())
    `

**Comment by krocheck** [comment]
So is [this change](https://github.com/bitfocus/companion-module-globalcache-itac-ir/commit/8816449aa7ed92ec2f0096f882c1b2726b442249) from a bit ago a factor?

**Comment by jpooley-hearst** [comment]
@krocheck there's been multiple changes made to get things working this year. Like I said, once we remove the erroneous comma from that line we should be set other than hitting the docs.

**Comment by jpooley-hearst** [comment]
Is anyone able to get the IP2IR module to even connect when in a Linux development environment? The current release hooks up just fine but I get null status with nothing useful showing up in console. I also can't get regex to parse correctly on the dev environment when it's working on Windows with a release build.

**Comment by sydneyjd** [comment]
I would love to help troubleshoot more if you would let me know how to get plugged in to do so. I am running Companion on Windows 10. I tried the Linux Appimage but it will not load for some reason.

**Comment by sydneyjd** [comment]
I managed to compile Companion on my Linux machine using the stable-2.1 branch. 
 If I remove `/((?:\d+(?:,)?)+)/g` from the `regex:''` string on line 114 this plugin works fine with my IR string on a Itac Ethernet, so long as I remove the `sendir,1:1,1,` section of the IR string. I did a bit of playing around trying to create a regex that would work, but I do not understand regular expressions well enough to be able to fix it. Any suggestions? Am I looking in the wrong spot?

**Comment by jpooley-hearst** [comment]
@sydneyjd I tried to replicate but I could not get the stable-2.1 branch to run on node 12.19.1 and 8.17.0 "HID.node was compiled against 72, this version requires 57". 

If anyone experienced with the underlying frameworks could chime in please do otherwise I'm at an impasse.

**Comment by sydneyjd** [comment]
I dug up an old commit which if I use the original regex string, everything seems to work fine. https://github.com/bitfocus/companion-module-globalcache-itac-ir/commit/2b4edd76e76cff881a6f90872466a68fbe5a81fa
I just published pretty much the same commit to my fork of this repo. Can anyone test this change? I can open a PR if someone wants. https://github.com/sydneyjd/companion-module-globalcache-itac-ir/commit/94da959ce9e8fb3afb4f2d75a5a9c10020e00973

As for the compile issue, I don't know what to say, and am not great at troubleshooting such things. I just compiled as to the instructions listed in DEVELOPER.md file. Running Debian 10.

**Comment by jpooley-hearst** [comment]
That will get us back to working, but that commit assumes that IR strings begin with 1 and that's not always the case. I can open an issue for that separately. 

We should also tighten up the regex as a separate issue. 

I can't get a Debian 10 Gnome setup working... I'm getting a Gtk warning cannot open display. 

Unable to test currently unless we can get a Windows build made.

**Comment by sydneyjd** [comment]
Yeah, I figured it was a bit hack-ish. Me just playing around not knowing exacly what im doing... learning I suppose.

I did not even try running the companion gui, I just invoked ./headless.js and was using the webUI.

**Comment by jpooley-hearst** [comment]
Ahh thanks and thanks for your work keeping the ship moving. 

I am still getting a null connection status from Debian 10 so I am unable to test

**Comment by ethan-tripp** [comment]
Happy to test this latest commit on my IP2IR devices when ready. Will it be a part of the next beta download? Thank you all for your work on this!

**Comment by sydneyjd** [comment]
At this point, what needs to be done before a fix can be pushed? I just don't want to have this get stagnant, as we plan to roll out a bunch of Itach devices over the next couple months, and really would like Companion compatibility. I can test code if you have a fix that is preferred.

**Comment by jpooley-hearst** [comment]
@JeffreyDavidsz could you effect an experimental build please?

**Comment by JeffreyDavidsz** [comment]
As far as I can see, latest commit is in the core. @sydneyjd did you create a pull request?

**Comment by sydneyjd** [comment]
@JeffreyDavidsz My bad. Just made a pull request. Thanks.

**Comment by sydneyjd** [comment]
@ethan-tripp Should be working in the latest beta build, just installed it this morning and all seems well. I will close this unless someone else has something to add.

**Comment by lfrisk** [comment]
I am experiencing the same issues that the initial post described:

- Companion: 2.1.2 Dec. 6th Build
- Global Cache ip2ir

Symptoms:

- iLearn app functions correctly when learning codes, and executing codes. This tells me that the hardware (ip2ir) device is configured correctly.
- When attempting to transfer the learned code to Companion 2.1.2 Dec. 06 build: the ip2ir led indicator flashes to indicate that it is receiving communication when I press the "Test Actions" button, but the device does not respond in the way that the same code worked on the iLearn software. 

![Screen Shot 2021-03-02 at 11 59 52 AM](https://user-images.githubusercontent.com/74016050/109693338-28148b80-7b4f-11eb-9f41-be13b77a8f44.png)

Code Sample: sendir,1:3,1,38109,1,1,128,64,16,16,16,48,16,16,16,48,16,16,16,48,16,16,16,48,16,16,16,48,16,16,16,48,16,48,16,16,16,48,16,16,16,48,16,48,16,48,16,48,16,16,16,48,16,16,16,16,16,16,16,48,16,48,16,16,16,48,16,16,16,16,16,16,16,16,16,16,16,16,16,16,16,48,16,16,16,16,16,16,16,48,16,16,16,16,16,16,16,48,16,16,16,48,16,16,16,3800

_NOTE: I attempted to follow the most current notes for configuring this device, and they said to remove the "sendir,1:3,1," from the code, and after reading this thread I see that this is no longer valid to do this._  

Attempting to control a Sony PMW EX30 Video Recording Deck using the IR blaster.

**Comment by sydneyjd** [comment]
This was fixed about a month ago and is not in the latest stable, as you have noticed. You can try one of the latest beta builds, or wait for the next stable release if you wish. You do still have to remove the "sendir,1:1,1," section as you mentioned. The original issue was that that sendir section was being classified as invalid, but did actually need to be removed.

**Comment by lfrisk** [comment]
Looks like I will have to try a beta build, currently if I remove the "sendir,1:1,1," section it registers as invalid.

Regards,
Larry Frisk | Technical Director
630-962-7834 | lfrisk@theorchardefc.org<mailto:lfrisk@theorchardefc.org>

THE ORCHARD
1301 South Grove Avenue, Barrington, IL 60010
theorchardefc.org<http://www.theorchardefc.org> | Facebook<https://www.facebook.com/TheOrchardEvangelicalFreeChurch/> | Instagram<https://www.instagram.com/theorchardefc/> | Twitter<https://twitter.com/TheOrchardEFC>

**Comment by lfrisk** [comment]
I wound up loading it as a generic tcp device. this is working and is stable.


Larry Frisk | Technical Director
630-962-7834 | lfrisk@theorchardefc.org<mailto:lfrisk@theorchardefc.org>

The Orchard Evangelical Free Church
1301 South Grove Avenue, Barrington, IL 60010

theorchardefc.org<https://nam12.safelinks.protection.outlook.com/?url=http%3A%2F%2Fwww.theorchardefc.org%2F&data=02%7C01%7Clfrisk%40theorchardefc.org%7C260a3f6913da431ad61908d869fbe097%7Ce9901c341da34636a30b6ebf7879e64b%7C0%7C0%7C637375877834155971&sdata=Yh0BItN9Ti325KBYxdmS5TCCUCA7Nisc%2FMe6PVK5OUY%3D&reserved=0> | Facebook<https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Fwww.facebook.com%2FTheOrchardEvangelicalFreeChurch&data=02%7C01%7Clfrisk%40theorchardefc.org%7C260a3f6913da431ad61908d869fbe097%7Ce9901c341da34636a30b6ebf7879e64b%7C0%7C0%7C637375877834165928&sdata=GRce3NFi1s2rxqybsJ6XU0ZmJuGUiKXRC8mzlCGOVqk%3D&reserved=0> | Twitter<https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Ftwitter.com%2Ftheorchardefc&data=02%7C01%7Clfrisk%40theorchardefc.org%7C260a3f6913da431ad61908d869fbe097%7Ce9901c341da34636a30b6ebf7879e64b%7C0%7C0%7C637375877834165928&sdata=yQJ1tWt%2BLUyWynvcLp357lHtLgQPHoMNxWtYz78hFwM%3D&reserved=0> | Instagram<https://nam12.safelinks.protection.outlook.com/?url=https%3A%2F%2Fwww.instagram.com%2Ftheorchardefc%2F&data=02%7C01%7Clfrisk%40theorchardefc.org%7C260a3f6913da431ad61908d869fbe097%7Ce9901c341da34636a30b6ebf7879e64b%7C0%7C0%7C637375877834165928&sdata=Z6pxuJy7YFZc1B1jcF2mSgRJIlgqBDf89QjPEy%2FVbIU%3D&reserved=0>

________________________________

**Comment by sydneyjd** [comment]
Good to hear you hacked up a solution. That is the other way to work around it for the time being.

**Comment by obcocav** [comment]
Should this be working in all the latest beta's?  It worked perfect on the 2792 build that listed the fix but hasn't worked for me in any beta since.  Do these fixes stick from version to version while in beta?  Thanks

**Comment by obcocav** [comment]
@sydneyjd.  Just discovered that this is still fixed in latest MAC beta but NOT in PC beta.  It does work in the 3117 MAC build.  Went back to the 2792 PC build and can't even get that one to work now.  So no build I have works on PC.  Hope this helps get a fix!

Thanks!

**Comment by obcocav** [comment]
@lfrisk .  Could you share the settings you used for the generic device to get it to work as a workaround?  Thanks!

**Comment by lfrisk** [comment]
1.  Create a new instance using “Generic TCP and UDP Requests”
     *   Set the ip address…
     *   not sure if there is a specific port that is needed, but mine was set for 4998
     *   Select TCP
  2.  Using the iLearn app, discover your remote command codes that you need for your devices
  3.  Button Programming:
     *   Select the only option “<device name of your choice> tcp-udp: Send Command”
     *   Command: paste the full Global Cache remote code (i.e. sendir,1:1,1,40064,3,1,97,24,24,24,24,24,24,24,48,24,48,24,24,24,24,24,24,24,48,24,24,24,48,24,48,24,24,24,24,24,48,24,48,24,24,24,48,24,24,24,24,557)
     *   Command End Character: select: “CRLF - \r\n (Common Windows)”

                                                              i.      I found that the first selection, “LF Mac” end character does not respond with the Global Cache unit.



I had to tweak these settings to get the devices to respond correctly:

  *   Ir port number
  *   Repeat settings

Drawbacks:

  *   This is a 1-way setup, you cannot build in any feedback from the device.

**Comment by sydneyjd** [comment]
@obcocav Just tested version 3031 for Windows, working perfectly for me. Not sure if you have something else going on in your setup. Does the IR code window accept the IR command minus the sendir,1:1,1 section?
